org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/TextEditUtils.java

```diff
@@ -0,0 +1,186 @@
+	 * given range (from, to) of the given text document and null otherwise.
+	 * 
+	 * @param from            the range from.
+	 * @param to              the range to.
+	 * @param expectedContent the expected content.
+	 * @param textDocument    the text document.
+	 * 
+	 * @return the {@link TextEdit} to insert the given expected content from the
+	 *         given range (from, to) of the given text document and null otherwise.
+	 */
+	public static TextEdit createTextEditIfNeeded(int from, int to, String expectedContent, TextDocument textDocument) {
+		String text = textDocument.getText();
+
+		// Check if content from the range [from, to] is the same than expected content
+		if (isMatchExpectedContent(from, to, expectedContent, text)) {			
+			// The expected content exists, no need to create a TextEdit
+			return null;
+		}
+
+		int i = expectedContent.length() - 1;
+		boolean matchExpectedContent = true;
+		while (from > 0) {
+			char c = text.charAt(from);
+			if (Character.isWhitespace(c)) {
+				if (matchExpectedContent) {
+					if (i < 0) {
+						matchExpectedContent = false;
+					} else {
+						if (expectedContent.charAt(i) != c) {
+							matchExpectedContent = false;
+						}
+						i--;
+					}
+				}
+			} else {
+				break;
+			}
+			from--;
+		}
+		from++;
+		if (matchExpectedContent) {
+			matchExpectedContent = to - from == expectedContent.length();
+		}
+
+		if (!matchExpectedContent) {
+			try {
+				Position endPos = textDocument.positionAt(to);
+				Position startPos = to == from ? endPos : textDocument.positionAt(from);
+				Range range = new Range(startPos, endPos);
+				return new TextEdit(range, expectedContent);
+			} catch (BadLocationException e) {
+				LOGGER.log(Level.SEVERE, e.getMessage(), e);
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Returns true if the given content from the range [from, to] of the given text
+	 * is the same than expected content and false otherwise.
+	 * 
+	 * @param from            the from range.
+	 * @param to              the to range.
+	 * @param expectedContent the expected content.
+	 * @param text            the text document.
+	 * 
+	 * @return true if the given content from the range [from, to] of the given text
+	 *         is the same than expected content and false otherwise.
+	 */
+	private static boolean isMatchExpectedContent(int from, int to, String expectedContent, String text) {
+		if (expectedContent.length() == to - from) {
+			int j = 0;
+			for (int i = from; i < to; i++) {
+				char c = text.charAt(i);
+				if (expectedContent.charAt(j) != c) {
+					return false;
+				}
+				j++;
+			}
+		} else {
+			return false;
+		}
+		return true;
+	}
+
+	public static String applyEdits(TextDocument document, List<? extends TextEdit> edits) throws BadLocationException {
+		String text = document.getText();
+		List<? extends TextEdit> sortedEdits = mergeSort(edits /* .map(getWellformedEdit) */, (a, b) -> {

```

Comment:
Is there any reason for using the custom merge sort instead of a built-in sorting function?

Reply:
I use now Comparator, but please not ethis method is used only for tests.